cfpb/cfgov-refresh
# Task per environment

If we only ran gulp tasks in the environments they are needed we would have a number of dependencies that wouldn't need to be installed. This PR adds the architecture to ./frontend.sh and gulp to allow segregating tasks to specific node environments.

## Removals

- `require-dir` dependency.

## Changes

- Moves dev dependencies that don't run in production environment to devDependencies in package.json.
- Updates gulpfile to introduce logic to only load tasks files that are called (except in top-level build, test, and default tasks).
- Updates gulpfile to support ignoring tasks from build based on environment.
- Updates `frontend.sh` to support rebuilding node_modules when switching environments.

## Testing

Testing development:
1. Run `./frontend.sh` (or `./frontend.sh development`, which is the default).
2. The npm deps are installed in node_modules, and a file `node_modules/NODE_ENV` is added, which will have the contents `development`.
3. Run `./frontend.sh` again. The npm modules should NOT install again and use the existing folder ("Dependencies are up to date." should be in the console log).
7. Run `echo $NODE_ENV`. It should output "development".
8. Run `gulp`. It should list all the gulp tasks in `gulp/tasks/` in the format `Requiring task ./gulp/tasks/[taskname].js`.
9. Run `gulp build`. It should show the same required tasks as above.
10. Run `gulp lint`. It should ONLY list `Requiring task ./gulp/tasks/lint.js` at the top of the output.
11. Run `gulp test:unit`. It should ONLY list `Requiring task ./gulp/tasks/test-unit.js` at the top of the output.

Testing gulp:
`gulp --tasks` outputs the tasks that can be run.
1. `gulp` should load and run all task files.
2. `gulp build` should load and run all task files (or a smaller subset when in production environment).
3. `gulp lint` should load only `lint.js` task.
4. `gulp test` should load all task files and run tests.
5. `gulp test:unit` should load only `test-unit.js` task.
6. `gulp test:unit:scripts` should load only `test-unit.js` task.

Testing production:
1. Run `./frontend.sh production`
2. The npm deps are installed in node_modules and a file `node_modules/NODE_ENV` is added, which will have the contents `production`.
3. Run `./frontend.sh production` again. The npm modules should NOT install again and use the existing folder.
4. Run `echo $NODE_ENV`. It should output "production".
5. Run `gulp`. It should list all the gulp tasks in `gulp/tasks/` in the format `Requiring task ./gulp/tasks/[taskname].js` EXCEPT audit, docs, lint, test-unit.
6. Run `gulp build`. It should show the same required tasks as above.
7. Run `gulp styles`. It should ONLY list `Requiring task ./gulp/tasks/styles.js` at the top of the output.
8. Run `gulp scripts:apps`. It should ONLY list `Requiring task ./gulp/tasks/scripts.js` at the top of the output.

## Notes

- For easier review, note that the test gulp tasks are unchanged, they are just split into two files to separate the unit and acceptance test code.

- I removed the `$NODE_DIR` variable from `frontend.sh`. It doesn't seem realistic that we wouldn't be using node_modules for this, so why not reference the path directly.

- Top-level tasks are in gulpfile.js. These tasks don't define their own tasks but instead call other tasks. We had `test.js` calling the lint task, which I don't think is right. Each task file should only be calling its own tasks, therefore this PR moves test to be top-level. The consequence of this is it acts like build and the default task in that it loads all gulp tasks. This is no worse than how it was before, it just means the test task doesn't utilize the optimizations that test:unit or test:acceptance use.
